### PR TITLE
WT-9851 Support bounded cursor next and prev traversal in FLCS

### PR DESCRIPTION
**Pull request by luke-pearson**

### CODE REVIEW

File: src/btree/bt_cursor.c [hunk 1]
```diff
@@ -300,21 +300,18 @@ __wt_cursor_valid(WT_CURSOR_BTREE *cbt, WT_ITEM *key, uint64_t recno, bool *vali
...
-                WT_RET(__btcur_bounds_contains_key(
-                  session, &cbt->iface, key, cbt->recno, &key_out_of_bounds, NULL));
-
-            /* The key value pair we were trying to return weren't within the given bounds. */
-            if (key_out_of_bounds)
-                return (0);
+            tmp_key.data = WT_INSERT_KEY(cbt->ins);
+            tmp_key.size = WT_INSERT_KEY_SIZE(cbt->ins);
         }
+        WT_RET(__btcur_bounds_contains_key(session, &cbt->iface, key == NULL ? &tmp_key : key,
```
Comment (jiechenbo) [inline src/btree/bt_cursor.c:308]: I would add a comment here to explain all the different possible scenarios that can happen here. E.g. cbt->recno, insert list and a self provided key that enter this function.

File: src/btree/bt_cursor.c [hunk 2]
```diff
@@ -300,21 +300,18 @@ __wt_cursor_valid(WT_CURSOR_BTREE *cbt, WT_ITEM *key, uint64_t recno, bool *vali
      * update that's been deleted is not a valid key/value pair).
      */
     if (cbt->ins != NULL) {
-        if (WT_CURSOR_BOUNDS_SET(&cbt->iface)) {
+        if (btree->type == BTREE_ROW && WT_CURSOR_BOUNDS_SET(&cbt->iface) && key == NULL) {
```
Comment (clarissecheah) [inline src/btree/bt_cursor.c:303]: Why do we no longer check `WT_CURSOR_BOUNDS_SET(&cbt->iface)` before executing any of lines 303-314? Is this to clean up the if statement here?

File: src/btree/bt_cursor.c [hunk 3]
```diff
@@ -324,6 +321,14 @@ __wt_cursor_valid(WT_CURSOR_BTREE *cbt, WT_ITEM *key, uint64_t recno, bool *vali
         }
     }
 
+    if (btree->type != BTREE_ROW) {
```
Comment (jiechenbo) [inline src/btree/bt_cursor.c:324]: Add comment here to explain that we are checking against the update list now.

**Review by jiechenbo — APPROVED** [review]
LGTM! I would just add a few comments.

**Review by clarissecheah — APPROVED** [review]
LGTM!

**Comment by luke-pearson** [comment]
@sauclovian-wt I added you as a reviewer here as you recently worked on the FLCS Logic within wiredtiger. The bounded cursor work may be a bit unfamiliar to you. If you need more context let me know and I will reach out via email.

File: src/btree/bt_cursor.c (hunk 1 above)
Comment (luke-pearson) [reply src/btree/bt_cursor.c:314]: 0f4f0221d

File: src/btree/bt_cursor.c (hunk 3 above)
Comment (luke-pearson) [reply src/btree/bt_cursor.c:335]: 0f4f0221d

**Review by jiechenbo — APPROVED** [review]
LGTM!

**Comment by sauclovian-wt** [comment]
Sorry I didn't get to this in time -- it looks ok but I have indeed not been following the bounds stuff, so my opinion probably isn't worth anything anyway.

**Comment by sauclovian-wt** [comment]
...as evidenced by it apparently not being ok and getting reverted :-|

**Comment by luke-pearson** [comment]
Yeah we'll look into it, I suspect its an edge case with respect to implicit keys or something.

**Pull request merged**